vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+  dst->xbutton.serial = XLastKnownRequestProcessed(fl_display);
+  dst->xbutton.display = fl_display;
+  dst->xbutton.window = wnd;
+  dst->xbutton.root = root;
+  dst->xbutton.subwindow = None;
+  dst->xbutton.time = fl_event_time;
+  dst->xbutton.x = src.event_x;
+  dst->xbutton.y = src.event_y;
+  dst->xbutton.x_root = x_root;
+  dst->xbutton.y_root = y_root;
+  dst->xbutton.state = state.mods;
+  dst->xbutton.state |= ((state.ptr_buttons >> 1) & 0x1f) << 8; // FIXME: Check this!
+  dst->xbutton.same_screen = True; // FIXME
+}
+
+void FLTKGestureHandler::fakeMotionEvent(const XIDeviceEvent* origEvent)
+{
+  XEvent fakeEvent;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = MotionNotify;
+  fakeEvent.xmotion.is_hint = False;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeMotionEvent(const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+  int mask;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = MotionNotify;
+  fakeEvent.xmotion.is_hint = False;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+  
+  switch (origEvent.gesture) {
+    case GH_LEFTBTN:
+      mask = Button1Mask;
+      break;
+    case GH_MIDDLEBTN:
+      mask = Button2Mask;
+      break;
+    case GH_RIGHTBTN:
+      mask = Button3Mask;
+      break;
+    //case GH_ZOOM:
+    //  mask = ControlMask;
+    //  break;
+    default:
+      mask = 0;
+  }
+
+  fakeEvent.xbutton.state |= mask;
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeButtonEvent(bool press, int button,
+                                         const XIDeviceEvent* origEvent)
+{
+  XEvent fakeEvent;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+
+  vlog.info("CORE xbutton.state = %i", fakeEvent.xbutton.state);
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeButtonEvent(bool press, int button,
+                                         const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+  int mask;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  prepXEvent(&fakeEvent, origEvent);
+
+  switch (button) {
+    case Button1:
+      mask = Button1Mask;
+      break;
+    case Button2:
+      mask = Button2Mask;
+      break;
+    case Button3:
+      mask = Button3Mask;
+      break;
+    case Button4:
+      mask = Button4Mask;
+      break;
+    case Button5:
+      mask = Button5Mask;
+      break;
+    case Button6:
+      mask = Button6Mask;
+      break;
+    case Button7:
+      mask = Button7Mask;
+      break;
+    default:
+      mask = 0;
+  }
+
+  if (press)
+    fakeEvent.xbutton.state |= mask;
+  else
+    fakeEvent.xbutton.state &= ~mask;
```

Comment:
Not sure we want to clear the bit in the off case that it is also pressed by another device.

Reply:
This was taken care of with 461988b21e9d7e5. According to my testing with `xev`, we should actually be setting the mask on button release events only.